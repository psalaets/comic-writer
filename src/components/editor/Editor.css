.Editor {
  flex: 0 0 50%;
  border-right: 1px solid #eeeeee;
  background-color: #fafafa;
}

.Editor__textarea {
  padding-left: 1em;
  padding-right: 1em;
  border: 0;
  outline: none;
  font-family: monospace;
  color: #232323;
  resize: none;
  width: 100%;
  height: 100%;
  font-size: 1em;
  background-color: transparent;
.Editor__scrollpast {
  height: 25vh;
}
