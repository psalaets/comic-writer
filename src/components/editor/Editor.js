import React, { Component } from 'react';
import PropTypes from 'prop-types';

import CodeMirror from './CodeMirror';

export default class Editor extends Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  render() {
    return (
      <CodeMirror
          editorWidth={80}
          value={this.props.value}
          cursor={this.props.cursor}
          stats={this.props.stats}
          onChange={this.handleChange}
        />
    )
  }

  handleChange(event) {
    this.props.onChange({
      value: event.value,
      cursor: event.cursor
    });
  }
}

Editor.propTypes = {
  cursor: PropTypes.shape({
    line: PropTypes.number.isRequired,
    ch: PropTypes.number.isRequired
  }).isRequired,
  value: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  stats: PropTypes.array.isRequired
};
