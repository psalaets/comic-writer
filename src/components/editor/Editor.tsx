import React, { Component } from 'react';
import { ComicStats } from '../../stats/types';
import CodeMirror from './CodeMirror';

type Props = {
  value: string;
  stats: Array<ComicStats>;
  characters: Array<string>;
  onChange: (event: EditorChangeEvent) => void;
}

export interface EditorChangeEvent {
  value: string;
}

export default class Editor extends Component<Props> {
  constructor(props: Props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  render() {
    return (
      <CodeMirror
        editorWidth={80}
        value={this.props.value}
        stats={this.props.stats}
        characters={this.props.characters}
        onChange={this.handleChange}
      />
    )
  }

  handleChange(event: EditorChangeEvent): void {
    this.props.onChange(event);
  }
}
