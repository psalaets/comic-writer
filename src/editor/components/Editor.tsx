import React, { Component } from 'react';

import { PanelCount, WordCount, EditorChangeEvent } from '../types';
import CodeMirror from './codemirror/CodeMirror';

type Props = {
  value: string,
  panelCounts: Array<PanelCount>,
  wordCounts: Array<WordCount>,
  characters: Array<string>,
  onChange: (event: EditorChangeEvent) => void
}

export default class Editor extends Component<Props> {
  constructor(props: Props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
  }

  render() {
    return (
      <CodeMirror
        editorWidth={80}
        value={this.props.value}
        panelCounts={this.props.panelCounts}
        wordCounts={this.props.wordCounts}
        characters={this.props.characters}
        onChange={this.handleChange}
      />
    )
  }

  handleChange(event: EditorChangeEvent): void {
    this.props.onChange(event);
  }
}
