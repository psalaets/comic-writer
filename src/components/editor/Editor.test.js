import React from 'react';
import renderer from 'react-test-renderer';

import Editor from './Editor';

describe('Editor', () => {
  describe('page markers', () => {
    it('first marker', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = `## `
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding marker at start', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## \n## Page 1\npage 1 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding marker at end', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n## '
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding marker in the middle', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n## \n## Page 2\npage 2 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('pasting page at start', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 3\npage 3 info\n## Page 1\npage 1 info\n## Page 2\npage 2 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('pasting page at end', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 2\npage 2 info\n## Page 1\npage 1 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('pasting page in middle', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n## Page 3\npage 3 info\n## Page 2\npage 2 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('remove page from start', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 2\npage 2 info\n## Page 3\npage 3 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('remove page from middle', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n## Page 3\npage 3 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })
  })

  describe('panel markers', () => {
    it('adding panel to page', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n### '
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding second panel to page', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n### Panel 1\n### '
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding panel at start', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n### \n### Panel 1\npanel 1 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding panel in middle', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\n### Panel 1\npanel 1 info\n### \n### Panel 2\npanel 2 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding panel to second page', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\n### Panel 1\npanel 1 info\n## Page 2\n### '
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('adding another panel to second page', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\n### Panel 1\npanel 1 info\n## Page 2\n### Panel 1\n### '
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('pasting panel at start', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\npage 1 info\n### Panel 2\n### Panel 1\npanel 1 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('pasting panel in middle', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\n### Panel 1\npanel 1 info\n### Panel 3\n### Panel 2\npanel 2 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('removing panel from start', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\n### Panel 2\npanel 2 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })

    it('removing panel from middle', () => {
      const component = renderer.create(<Editor />);
      let tree = component.toJSON();

      const value = '## Page 1\n### Panel 1\npanel 1 info\n### Panel 3\npanel 3 info'
      tree.props.onChange(onChangeEvent(value));

      tree = component.toJSON();

      expect(tree).toMatchSnapshot();
    })
  })

  describe('onChange event', () => {
    describe('value', () => {
      it('when changed with basic text', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        const value = 'this is some basic text'
        tree.props.onChange(onChangeEvent(value));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].value).toBe('this is some basic text');
      })

      it('changed with markdown', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        const value = '## '
        tree.props.onChange(onChangeEvent(value));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].value).toBe('## Page 1');
      })

      it('cleared', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        tree.props.onChange(onChangeEvent(''));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].value).toBe('');
      })
    })

    describe('sourceLength', () => {
      it('empty value', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        tree.props.onChange(onChangeEvent(''));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].sourceLength).toBe(0);
      })

      it('single line', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        tree.props.onChange(onChangeEvent('blah'));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].sourceLength).toBe(4);
      })

      it('multiple lines', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        tree.props.onChange(onChangeEvent('## Page 1\n\nblah'));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].sourceLength).toBe(15);
      })
    })

    describe('cursorAtEnd', () => {
      it('cursor at the end of value', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        tree.props.onChange(onChangeEvent('abc', 3));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].cursorAtEnd).toBe(true);
      })

      it('cursor in middle of value', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        tree.props.onChange(onChangeEvent('abc', 1));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].cursorAtEnd).toBe(false);
      })
    })

    describe('cursorPage and cursorPanel', () => {
      it('cursor in a page and panel', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        const value = '## Page 1\n## Page 2\n### Panel 1\nadsf';
        tree.props.onChange(onChangeEvent(value, value.length - 3));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].cursorPage).toBe(2);
        expect(onChange.mock.calls[0][0].cursorPanel).toBe(1);
      })

      it('cursor has no page nor panel', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        const value = 'asdf';
        tree.props.onChange(onChangeEvent(value, value.length - 3));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].cursorPage).toBe(undefined);
        expect(onChange.mock.calls[0][0].cursorPanel).toBe(undefined);
      })

      it('cursor has page only', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        const value = '## Page 1\n## Page 2\nasdf';
        tree.props.onChange(onChangeEvent(value, value.length - 3));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].cursorPage).toBe(2);
        expect(onChange.mock.calls[0][0].cursorPanel).toBe(undefined);
      })

      it('cursor has panel only', () => {
        const onChange = jest.fn();
        const component = renderer.create(<Editor onChange={onChange} />);
        let tree = component.toJSON();

        const value = '### Panel 1\n### Panel 2\nasdf';
        tree.props.onChange(onChangeEvent(value, value.length - 3));

        expect(onChange.mock.calls.length).toBe(1);
        expect(onChange.mock.calls[0][0].cursorPage).toBe(undefined);
        expect(onChange.mock.calls[0][0].cursorPanel).toBe(2);
      })
    })
  })
})

function onChangeEvent(text, selectionStart = 0) {
  return {
    target: {
      value: text,
      selectionStart
    }
  };
}