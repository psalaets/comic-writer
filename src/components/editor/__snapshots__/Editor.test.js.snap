// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor page markers adding marker at end 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
## Page 2"
/>
`;

exports[`Editor page markers adding marker at start 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
## Page 2
page 1 info"
/>
`;

exports[`Editor page markers adding marker in the middle 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
## Page 2
## Page 3
page 2 info"
/>
`;

exports[`Editor page markers first marker 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1"
/>
`;

exports[`Editor page markers pasting content with page count in double digits 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
## Page 2
## Page 3
## Page 4
## Page 5
## Page 6
## Page 7
## Page 8
## Page 9
## Page 10
## Page 11
## Page 12
## Page 13"
/>
`;

exports[`Editor page markers pasting page at end 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 2 info
## Page 2
page 1 info"
/>
`;

exports[`Editor page markers pasting page at start 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 3 info
## Page 2
page 1 info
## Page 3
page 2 info"
/>
`;

exports[`Editor page markers pasting page in middle 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
## Page 2
page 3 info
## Page 3
page 2 info"
/>
`;

exports[`Editor page markers remove page from middle 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
## Page 2
page 3 info"
/>
`;

exports[`Editor page markers remove page from start 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 2 info
## Page 2
page 3 info"
/>
`;

exports[`Editor panel markers adding another panel to second page 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
### Panel 1
panel 1 info
## Page 2
### Panel 1
### Panel 2"
/>
`;

exports[`Editor panel markers adding panel at start 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
### Panel 1
### Panel 2
panel 1 info"
/>
`;

exports[`Editor panel markers adding panel in middle 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
### Panel 1
panel 1 info
### Panel 2
### Panel 3
panel 2 info"
/>
`;

exports[`Editor panel markers adding panel to page 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
### Panel 1"
/>
`;

exports[`Editor panel markers adding panel to second page 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
### Panel 1
panel 1 info
## Page 2
### Panel 1"
/>
`;

exports[`Editor panel markers adding second panel to page 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
### Panel 1
### Panel 2"
/>
`;

exports[`Editor panel markers pasting panel at start 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
page 1 info
### Panel 1
### Panel 2
panel 1 info"
/>
`;

exports[`Editor panel markers pasting panel in middle 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
### Panel 1
panel 1 info
### Panel 2
### Panel 3
panel 2 info"
/>
`;

exports[`Editor panel markers removing panel from middle 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
### Panel 1
panel 1 info
### Panel 2
panel 3 info"
/>
`;

exports[`Editor panel markers removing panel from start 1`] = `
<textarea
  className="Editor"
  id="editor"
  onChange={[Function]}
  onScroll={[Function]}
  value="## Page 1
### Panel 1
panel 2 info"
/>
`;
